src/rules/configoption.ts

```diff
@@ -0,0 +1,217 @@
+import * as _ from "lodash";
+import { Predicate, RuleMatchedAt, YAMLRule } from "../lint";
+import { ConfigOption, ConfigSection } from "../replicated";
+import {
+  FoundValue,
+  ValueTraverser,
+  TraverseSearcher,
+  ValueSearcher,
+} from "../traverse";
+import { And, AnyOf, Match, Neq } from "../engine";
+
+class ConfigOptionExists implements Predicate<any> {
+  constructor(private readonly valueSearcher: ValueSearcher) {
+  }
+
+  public test(root: any): RuleMatchedAt {
+    const configItemNames = _.flatMap(root.config as ConfigSection[],
+      section => _.map(section.items,
+        item => item.name,
+      ),
+    );
+
+    const configOptionFinder = (v: any) => /{{repl ConfigOption/.test(v);
+    const configOptionUsages: FoundValue[] = this.valueSearcher.searchMatch(root, configOptionFinder);
+
+    // usages that contain NONE OF configItemNames
+    // every usage should match exactly one configItem
+    const filtered: FoundValue[] = _.filter(configOptionUsages, configOptionUsage => {
+      let matchesOne = false;
+      _.forEach(configItemNames, itemName => {
+        if (configOptionUsage.value.indexOf(itemName) !== -1) {
+          matchesOne = true;
+        }
+      });
+      return !matchesOne;
+    });
+
+    return { matched: !!filtered.length, paths: _.map(filtered, f => f.path) };
+  }
+}
+
+class ConfigOptionIsCircular implements Predicate<ConfigOption> {
+  constructor(private readonly valueSearcher: ValueSearcher) {
+  }
+
+  public test(configOption: ConfigOption): RuleMatchedAt {
+
+    const patternOption: string = `ConfigOption "${configOption.name}"`;
+    const patternOptionEquals: string = `ConfigOptionEquals "${configOption.name}"`;
+
+    const configOptionUsages: FoundValue[] = _.concat(
+      this.valueSearcher.searchContains(configOption, patternOption),
+      this.valueSearcher.searchContains(configOption, patternOptionEquals),
+    );
+
+    return {
+      matched: !!configOptionUsages.length,
+      paths: _.map(configOptionUsages, f => f.path),
+    };
+  }
+}
+
+export const configOptionExists: YAMLRule = {
+  name: "tmpl-configoption-exists",
+  type: "warn",
+  message: "Options referenced with `{{repl ConfigOption }}` must be present in the `config` section",
+  test: new ConfigOptionExists(new TraverseSearcher(new ValueTraverser())),
+
+  examples: {
+    wrong: [{
+      description: "Config Option `not_existent` is not defined in `config` section",
+      yaml: `
+---
+config:
+- name: hostname
```

Comment:
wanna call this `hostname_group` so its less confusing what youre validating here